public/app/plugins/panel/alertlist/unified-alerting/GroupedView.tsx

```diff
@@ -16,27 +16,36 @@ type Props = {
   options: UnifiedAlertListOptions;
 };
 
+export const UNGROUPED_KEY = '__ungrouped__';
+
 const GroupedModeView = ({ rules, options }: Props) => {
   const styles = useStyles2(getStyles);
-
   const groupBy = options.groupBy;
 
   const groupedRules = useMemo<GroupedRules>(() => {
     const groupedRules = new Map<string, Alert[]>();
 
-    const hasInstancesWithMatchingLabels = (rule: CombinedRuleWithLocation) =>
-      groupBy ? alertHasEveryLabelForCombinedRules(rule, groupBy) : true;
+    const hasInstancesWithMatchingLabels = (rule: CombinedRuleWithLocation) => {
+      return groupBy ? alertHasEveryLabelForCombinedRules(rule, groupBy) : true;
+    };
+
+    rules.forEach((rule) => {
+      const alertingRule = getAlertingRule(rule);
+      const hasInstancesMatching = hasInstancesWithMatchingLabels(rule);
 
-    const matchingRules = rules.filter(hasInstancesWithMatchingLabels);
-    matchingRules.forEach((rule: CombinedRuleWithLocation) => {
-      const alertingRule: AlertingRule | null = getAlertingRule(rule);
       (alertingRule?.alerts ?? []).forEach((alert) => {
-        const mapKey = createMapKey(groupBy, alert.labels);
+        const mapKey = hasInstancesMatching ? createMapKey(groupBy, alert.labels) : UNGROUPED_KEY;
+
         const existingAlerts = groupedRules.get(mapKey) ?? [];
         groupedRules.set(mapKey, [...existingAlerts, alert]);
       });
     });
 
+    // move the "UNGROUPED" key to the last item in the Map, items are shown in insertion order
+    const ungrouped = groupedRules.get(UNGROUPED_KEY) ?? [];
+    groupedRules.delete(UNGROUPED_KEY);
```

Comment:
why do you need to delete before setting the new value?

Reply:
If we don't delete the key first it will just overwrite the value but leave the place of the key intact (in other words, the key is already inserted so its order in the map isn't updated).

To achieve what we want here (which is to have the "ungrouped" group show up last) we have to delete the key and insert it again so its position in the map is update.

Reply:
got it! thank you for the explanation 👍